### PR TITLE
feat(sessions): Implement automatic session tracking

### LINKED ISSUE #1711: Session tracking/release health
Implement session tracking for the ruby SDK.

* spawn new thread, flush session stats every 60s
* add session abstraction, add as envelope items
* collect aggregate session stats in rack middleware

References
* https://develop.sentry.dev/sdk/sessions/
* https://docs.sentry.io/product/releases/health/

### PR DESCRIPTION
**Pull request by sl0thentr0py**
This adds automatic session tracking to our rack middleware (and hence rails).
The key components here are a `SessionFlusher` that spawns a new thread that sends pending aggregate sessions once a minute and a simple `Session` class for recording those individual sessions.

Note that we currently only implement [`request` mode sessions](https://develop.sentry.dev/sdk/sessions/#sdk-considerations) which are aggregated in the SDK as compared to `application` mode sessions which are sent as soon as they finish. The main reason for this is the scale of most servers out there which would overload our ingestion pipelines. 

The `Session` class could potentially be extended in the future for `application` mode sessions but for now it doesn't add much user value.

There are a couple of TODOs left which are slightly orthogonal to the main feature so I will make separate PRs for those.

closes #1711 

### References
* https://develop.sentry.dev/sdk/sessions/

---

### Implementation progress

- [x] Session flusher class, spawn new thread
- [x] Session class
- [x] Aggregate sessions
- [x] Hub / client apis
- [x] Rack middleware hook
- [x] Rails end-end
- [x] test coverage

In separate PRs
- [x] move rate limiting logic to envelope and unify `send_envelope` / `send_event`
- [x] ~exception handled / session crashed status~ moved to #1743

### CODE REVIEW

File: sentry-ruby/lib/sentry/transport.rb [hunk 1]
```diff
@@ -129,7 +135,7 @@ def encode(event)
       client_report_headers, client_report_payload = fetch_pending_client_report
       envelope.add_item(client_report_headers, client_report_payload) if client_report_headers
 
-      log_info("Sending envelope [#{item_type}] #{event_id} to Sentry")
+      log_info("[Transport] Sending envelope [#{item_type}] #{event_id} to Sentry")
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/transport.rb:138]: yes that was what I was planning in the separate envelope/rate limit PR.

File: sentry-ruby/lib/sentry/session_flusher.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      @pending_aggregates.map do |started, aggregates|
+        aggregates[:started] = started.iso8601
+        aggregates
+      end
+    end
+
+    def ensure_thread
+      return if @thread&.alive?
+
+      @thread = Thread.new do
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: yes I should have posted this in the PR description.
Basically I looked at using `TimerTask` because what we need in this thread is a 60 second scheduler.
I found [several issues](https://github.com/ruby-concurrency/concurrent-ruby/issues?q=TimerTask) and most importantly [this one regarding thread leaking]. But if we go with just a single `ThreadPoolWorker`, I think that should be fine and I can make those changes.

**Comment by sl0thentr0py** [comment]
@st0012 moved the envelope changes to #1742 and changed the base branch here.

File: sentry-ruby/lib/sentry/session_flusher.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: fwiw I did test the core thread logic with [this toy flusher](https://gist.github.com/sl0thentr0py/fb9558715cdebb642090ad638273fc41).

[the remark this answers is not in the document]
Comment (st0012) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: I still think `ThreadPoolExecutor` can also be used here? Something like:

```rb
executor.post do
  loop do
    sleep Flusher::FLUSH_INTERVAL
    flusher.flush
  end
end
```

The benefits are:

- We can terminate it the same way [we do for background worker](https://github.com/getsentry/sentry-ruby/blob/cf1e5b24788fb50b3ea49ecf9b1100162e1741e6/sentry-ruby/lib/sentry/background_worker.rb#L41-L44), which is more flexible then `kill`.
- Having a unified abstraction for all the background worker/thread is better for maintainability.

And the only cons is probably that we need to make some adjustments to the `BackgroundWorker` first to make some logics sharable. But I think that's fine and I can work on it too.

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: I don't know, I don't think `ThreadPoolExecutor` is meant for long-lived threads, it has 
```
    #   @option opts [Integer] :idletime (DEFAULT_THREAD_IDLETIMEOUT) the maximum
    #     number of seconds a thread may be idle before being reclaimed
```

We need a _scheduler_ thread that lives long and schedules the sending every 60s. The right abstraction for that is `TimerTask` which is unfortunately broken. `concurrent-ruby` has a lot of general abstractions with much more logic in each of them and I actually prefer this very simple implementation rather than shoe-horning logic into something that might not work.

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: we have the same simple logic in the [python SDK](https://github.com/getsentry/sentry-python/blob/b63b5ae32d254771cbd027009d0b6def7e2e07fb/sentry_sdk/sessions.py#L106-L126).

[the remark this answers is not in the document]
Comment (st0012) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: ok, let's give it a try. and just to confirm, it doesn't hang when the program exists?

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: cool, if you're fine with it, I'll do some extensive testing of the branch again on Monday for threading edge cases.

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: @st0012 quick note on the testing non-determinism, the way the python tests solve this is to have an explicit `flush` on the worker which gets rid of the timing problem by just waiting for everything to finish. Caveat there is that we can not test the threading/concurrent logic itself which is fairly tricky in general anyway.

**Comment by sl0thentr0py** [comment]
So I'm gonna document my testing here one by one.
The setup is as follows: 
I'm using [this branch](https://github.com/getsentry/sentry-ruby/compare/sessions...test-session) with a simple logging of the total thread counts and the session flusher thread count every 10 seconds and on init and exit. And the rails app I'm using is [here](https://github.com/sl0thentr0py/sentry_sample_apps/tree/main/ruby/sentry_rails_sample). 

First run, normal behavior with some success and error routes, thread counts are 
```
INIT STATE
--------------------
total count: 1
flusher count: 0
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 1
--------------------
EXIT STATE
--------------------
total count: 9
flusher count: 0
--------------------
```

and proof that end-end flow worked

image

**Review by antonpirker — APPROVED** [review]

File: sentry-ruby/lib/sentry/session.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Sentry
+  class Session
+    attr_reader :started, :status
+
+    # TODO-neel add :crashed after adding handled
+    STATUSES = %i(ok errored exited)
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session.rb:8]: nope

**Comment by antonpirker** [comment]
My review is just very on the surface, because I am not an Ruby export. But everything looks clean (and done the same way as in Python)

**Comment by sl0thentr0py** [comment]
thx for taking a look!

**Comment by sl0thentr0py** [comment]
Next test was to test the failure resiliency of the flusher. Basically checking two things
* if there's an exception inside the thread, the `ensure_thread` ensures that a new thread respawns eventually
* there is only a single flusher thread at a time, other threads exit/don't leak

This was done with [this commit](https://github.com/getsentry/sentry-ruby/commit/d56901e4ea258abf5b70009a5438238db4229418) to throw a random exception with 25% probability.

you can see this in the rails log when the exception occurs
```
#<Thread:0x00007f91d0123ce8@/Users/neel/sentry/sdks/sentry-ruby/sentry-ruby/lib/sentry/session_flusher.rb:77 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	3: from /Users/neel/sentry/sdks/sentry-ruby/sentry-ruby/lib/sentry/session_flusher.rb:78:in `block in ensure_thread'
	2: from /Users/neel/sentry/sdks/sentry-ruby/sentry-ruby/lib/sentry/session_flusher.rb:78:in `loop'
	1: from /Users/neel/sentry/sdks/sentry-ruby/sentry-ruby/lib/sentry/session_flusher.rb:80:in `block (2 levels) in ensure_thread'
/Users/neel/sentry/sdks/sentry-ruby/sentry-ruby/lib/sentry/session_flusher.rb:21:in `flush': Oops in thread (RuntimeError)
```

and the thread count log shows 0 for the flusher thread count when that happens. We see that the thread count will always be 0/1 and we will not add more threads in any way if there is already an existing thread.
```
INIT STATE
--------------------
total count: 1
flusher count: 0
--------------------
--------------------
total count: 26
flusher count: 1
--------------------
--------------------
total count: 26
flusher count: 1
--------------------
--------------------
total count: 25
flusher count: 0
--------------------
--------------------
total count: 25
flusher count: 0
--------------------
--------------------
total count: 26
flusher count: 1
--------------------
--------------------
total count: 26
flusher count: 1
--------------------
--------------------
total count: 26
flusher count: 1
--------------------
EXIT STATE
--------------------
total count: 9
flusher count: 0
--------------------
```

File: sentry-ruby/lib/sentry/session_flusher.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb:75]: @st0012 I've added my manual testing scenarios and proofs/logs below in separate comments. Let me know if you have any questions!

File: sentry-ruby/spec/sentry/session_flusher_spec.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      before { configuration.release = nil }
+
+      it "noops" do
+        flusher = described_class.new(configuration, client)
+        flusher.add_session(session)
+        expect(flusher.instance_variable_get(:@pending_aggregates)).to eq({})
+      end
+    end
+
+    it "spawns new thread" do
```
Comment (st0012) [inline sentry-ruby/spec/sentry/session_flusher_spec.rb:84]: Can we also add a case to check that it doesn't spawn another thread if there's already one?

**Review by st0012 — COMMENTED** [review]
I finally got time to do a more thorough review, sorry for the waiting.
The overall feature still looks good 👍 I only added some nitpick-ish naming/testing suggestions 🙂

File: sentry-ruby/lib/sentry/session.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session.rb:8]: might be used later on if we do application-mode sessions, I kept it in for self-documenting and specifying somewhere that `:ok` is not aggregated.

File: sentry-ruby/lib/sentry/session.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    AGGREGATE_STATUSES = %i(errored exited)
+
+    def initialize
+      @started = Sentry.utc_now
+      @status = :ok
+    end
+
+    def update_from_event(event)
+      return unless event.is_a?(Event) && event.exception
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session.rb:17]: yea you're right, I did it this way for consistency with python but it just adds confusion, will change

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session.rb:17]: but I will keep `update_from_exception` since one #1743 is implemented, we will decide if the status is `:crashed` or `:errored` from the `mechanism` property.

File: sentry-ruby/lib/sentry/transport.rb [hunk 6]
```diff
@@ -67,6 +67,8 @@ def is_rate_limited?(item_type)
         case item_type
         when "transaction"
           @rate_limits["transaction"]
+        when "sessions"
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/transport.rb:70]: the match is on our `item_type` which is `sessions`, but the rate limit category is `session` for both envelopes of the type `session` (application mode/single) and `sessions` (request mode/aggregate).

File: sentry-ruby/spec/sentry/session_flusher_spec.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        subject.flush
+      end.not_to change { transport.envelopes.count }
+    end
+
+    it "captures pending_aggregates in background worker" do
+      t1 = Sentry.utc_now
+      t1_bucket = Time.utc(t1.year, t1.month, t1.day, t1.hour, t1.min)
+
+      aggregates = { t1_bucket => { exited: 50, errored: 10 } }
+      subject.instance_variable_set(:@pending_aggregates, aggregates)
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/spec/sentry/session_flusher_spec.rb:52]: Again I'm not sure how I should rspec properly, so..
There are already higher level integration tests in `capture_exceptions_spec`. This block is for testing the behaviour of `#flush`. If I initiate `Session` objects and add them via `add_session` in a `before` block, I'm testing more that just the `#flush` method, so not sure what the right way to approach this is.

File: sentry-ruby/spec/sentry/session_flusher_spec.rb (hunk 4 above)
Comment (sl0thentr0py) [reply sentry-ruby/spec/sentry/session_flusher_spec.rb:84]: done

**Comment by sl0thentr0py** [comment]
@st0012 made some updates and left comments for the rest

File: sentry-ruby/lib/sentry/transport.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (st0012) [reply sentry-ruby/lib/sentry/transport.rb:70]: I see, thanks 👍

File: sentry-ruby/lib/sentry/session_flusher.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      @pending_aggregates = {}
+    end
+
+    def add_session(session)
+      return unless @release
+
+      ensure_thread
+
+      return unless Session::AGGREGATE_STATUSES.include?(session.status)
+      @pending_aggregates[session.started_bucket] ||= init_aggregates
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/lib/sentry/session_flusher.rb]: ahh sorry I misread your comment :facepalm: ignore this I will make the change.

File: sentry-ruby/spec/sentry/session_flusher_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (st0012) [reply sentry-ruby/spec/sentry/session_flusher_spec.rb:52]: I think it's not related to the test framework and it's more like a personal opinion tbh.
Here are 2 options:
1. Set up tests with public APIs, which will involve other parts into the tests.
2. Set up tests with private APIs, which will bind the test with implementation details.

Generally I prefer 1) because when we refactor implementations, 2) usually requires us to change the tests as well. And that'll increase the chance of breaking things without knowing it.

The downside of 1) is that when the public API we use for test setup failed (`#add_session`), this test will break as well. The worst case of this is that we won't be sure about whether the cause is on `add_session` malfunction or it's caused by `flush`. But given that we also test `add_session`, it should be easy to distinguish.

[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry-ruby/spec/sentry/session_flusher_spec.rb]: done

**Comment by sl0thentr0py** [comment]
@st0012  done!

File: sentry-ruby/lib/sentry-ruby.rb [hunk 9]
```diff
@@ -189,11 +194,18 @@ def init(&block)
       @main_hub = hub
       @background_worker = Sentry::BackgroundWorker.new(config)
 
+      @session_flusher = if config.auto_session_tracking
+                           Sentry::SessionFlusher.new(config, client)
```
Comment (st0012) [inline sentry-ruby/lib/sentry-ruby.rb:198]: @sl0thentr0py I think we should also check if the SDK is deactivated before initializing the flusher, especially to avoid the additional thread that won't be used. We can use `Configuration#enabled_in_current_env?` for it.
